pkg/router/crl/crl.go

```diff
@@ -209,30 +388,111 @@ func getHTTPCRL(url string) (*pkix.CertificateList, error) {
+	//	1. mtlsLatestSymlink: the haproxy template refers to this, so it's needed to prevent reload failures
+	//	2. the directory that mtlsLatestSymlink points to (same reason as #1)
+	//	3. 'mtlsLatestStagingSymlink: this only exists in the middle of a CA bundle/CRL refresh, and will be dealt with
+	//	   by the CRL management loop
+	//	4. the newest directory (may be the same as #2): if mtlsLatestSymlink isn't pointing to this directory, we may
+	//	   still be in the middle of a CA bundle/CRL refresh, so the CRL management loop might be actively working with
+	//	   this directory
+	// Everything else can be deleted
+	var newestDirName string
+	for _, dirEntry := range dirEntries {
+		entryName := dirEntry.Name()
+		// entryName is the bare file/directory name, but mtlsLatestSymlink, mtlsLatestStagingSymlink, and inUseDirName
+		// are all fully-qualified paths. Since we know that all the dir entries are the files/directories in
+		// mtlsBaseDirectory, we can prepend that to entryName to get the full name for comparisons now, and for
+		// os.RemoveAll() if it needs to be deleted later in this function
+		fullPath := filepath.Join(mtlsBaseDirectory, entryName)
+		log.V(4).Info(fmt.Sprintf("Checking %s (%s)", entryName, fullPath))
+		// Leave mtlsLatestSymlink, mtlsLatestStagingSymlink, and the current directory pointed to by
+		// mtlsLatestSymlink alone.
+		if fullPath == mtlsLatestSymlink || fullPath == mtlsLatestStagingSymlink || fullPath == inUseDirName {
+			continue
+		}
+		dirEntryInfo, err := dirEntry.Info()
 		if err != nil {
-			return crlForKeyId, err
+			continue
 		}
-		for _, ext := range crl.TBSCertList.Extensions {
-			if ext.Id.Equal(authorityKeyIdentifierOID) {
-				var authKeyId authorityKeyIdentifier
-				if _, err := asn1.Unmarshal(ext.Value, &authKeyId); err != nil {
-					return crlForKeyId, err
+		dirUpdateTimestamp := dirEntryInfo.ModTime()
+		if dirUpdateTimestamp.Before(mostRecentTimestamp) {
+			// If this isn't the newest directory and it's not pointed to by mtlsLatestSymlink, we definitely don't need it.
+			log.V(2).Info(fmt.Sprintf("Removing %s because it's stale", entryName))
+			err := os.RemoveAll(filepath.Join(mtlsBaseDirectory, entryName))
+			if err != nil {
+				log.Error(err, fmt.Sprintf("Failed to remove %s", entryName))
+				continue
+			}
+		} else {
+			// If this is newer than the newest directory yet seen, delete the directory we thought was newest, then
+			// keep its name as the current newest directory
+			mostRecentTimestamp = dirUpdateTimestamp
+			if newestDirName != "" {
+				log.V(2).Info(fmt.Sprintf("Removing %s because it's stale", newestDirName))
+				err := os.RemoveAll(filepath.Join(mtlsBaseDirectory, newestDirName))
+				if err != nil {
+					log.Error(err, fmt.Sprintf("Failed to remove %s", newestDirName))
 				}
-				subjectKeyId := hex.EncodeToString(authKeyId.KeyIdentifier)
-				crlForKeyId[subjectKeyId] = crl
 			}
+			newestDirName = entryName
 		}
```

Comment:
Are you sure you want this `else` branch?  I thought the idea was to use the timestamp of `mtlsLatestSymlink`'s target (i.e. `mostRecentTimestamp`) as the boundary:

* Delete things older than `mostRecentTimestamp` (i.e. the timestamp of `mtlsLatestSymlink`'s target).
* Keep `mtlsLatestSymlink`, its target, and anything newer than `mostRecentTimestamp`.

Using `mostRecentTimestamp = mtlsLatestSymlink` as an immutable boundary makes it easy to reason about what gets deleted.  But now you're moving the boundary, which makes this code more difficult to reason about, and it isn't clear why you need to change the boundary while you're using it.  

In particular, it seems to me like you could end up in trouble if somehow you have multiple newer tmpdirs and the newest one isn't the next one that you'll use as the target of `mtls-latest`:

| filename          | mtime |
|-------------------|-------|
| `mtls-latest`     |     1 |
| `mtls-latest.new` |     2 |
| `tmpdir1`         |     4 |
| `tmpdir2`         |     3 |

When you process `tmpdir1`, you set `mostRecentTimestamp` to 4, and then when you process `tmpdir2`, you end up deleting it, but maybe `tmpdir2` was to be the next target of `mtls-latest`!  Is that not possible with this logic?